### DISCUSSION TITLE
nested submodules detection w/o .gitmodules file

### DISCUSSION THREAD

**Discussion by andry81** [gitextensions/gitextensions#10644] [Q&A]
### Feature description

There is a feature in the git, when you can declare a working copy as a submodule through the `.git/config` out of a specific git submodule command:

https://git-scm.com/docs/git-submodule#Documentation/git-submodule.txt-init--ltpathgt82308203

> Optional \<path> arguments limit which submodules will be initialized. If no path is specified and submodule.active has been configured, submodules configured to be active will be initialized, otherwise all submodules are initialized.

`.git/config` example:
```
[submodule]
	active = .
```

The GitExtensions does ignore that and not show the nested working copies as submodules.

But the `git submodule summary` command shows them.

The steps to reproduce exists as the issue of a wrapper tool using this feature:

https://github.com/chronoxor/gil/issues/11#issuecomment-1374750059

> **Note**
You need to use `gil clone` in `non administrative` console and avoid links creation (symbolic links will involve submodules recursion).
The `git clone ...` and the rest won't initialize the submodule because there is no `.gitmodules` file.

The suggested feature would be parse all nested directories to find the `.git/config` with the `[submodule]` section or use some git API to properly discover nested submodules.

### Environment

- Git Extensions 4.0.1.15887
- Build f2567dea28bbf0e2cbcd2a95d4c7200e47d0d671
- Git 2.39.0.windows.2
- Microsoft Windows NT 6.1.7601 Service Pack 1
- .NET 6.0.13
- DPI 96dpi (no scaling)
- Microsoft.WindowsDesktop.App Versions

```
    Microsoft.WindowsDesktop.App 6.0.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 6.0.13 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

**Comment by vbjay** [comment]
Adding active submodule doc ref here https://git-scm.com/docs/gitsubmodules#_active_submodules.  I think using . in active is a bad idea.  Read the docs and see what pathspec is supposed to do.

**Comment by vbjay** [comment]
Provide either a bash script to create a local repo or create sample repo(s) that provide the details of what is expected vs what you get.

**Comment by andry81** [comment]
> Provide either a bash script to create a local repo or create sample repo(s) that provide the details of what is expected vs what you get.

They already created in the referenced issue above:

https://github.com/andry81-tests/gil-test1-A
https://github.com/andry81-tests/gil-test1-B

If you don't like to use gil tool, then you can just use raw git calls:

```
* git clone https://github.com/andry81-tests/gil-test1-A X
* git clone https://github.com/andry81-tests/gil-test1-B Y
* Move *content* of the `Y` into `X/_externals/gil-test1-B` directory
```

Now:

```
...\X>git submodule summary
* _externals/gil-test1-B c03e991...5103fe5:
  Warn: _externals/gil-test1-B doesn't contain commit c03e9919f2d9fe9b4c2c42b86a64ddc5a9da40f9
```

But if run the GitExt browse on the `X`, then it won't show `X/_externals/gil-test1-B` as a submodule to `X`.

**Comment by vbjay** [comment]
Submodule summary may list nested repos but when you list submodules by running ```git submodule```
![image](https://user-images.githubusercontent.com/3629489/212487916-df394de1-5a05-4968-9ffd-e76c0a2de041.png)

You are thinking it is a submodule when it is not.  git submodule results in

![image](https://user-images.githubusercontent.com/3629489/212488447-df3dcfc9-90f9-437e-9032-61b1313e11c1.png)

git diff does detect it as a nested repo 
![image](https://user-images.githubusercontent.com/3629489/212488522-cb167f08-4735-4577-90f7-dacaf197b948.png)

and git status 
![image](https://user-images.githubusercontent.com/3629489/212488542-f360b410-6ebd-48b2-a76c-bba46856ddbb.png)


You have a funky state and I doubt is supported as valid git state.  

So just because git is detecting nested repo and helping you out in summary call, it is not a submodule.  The modules file is required to actually work git submodules.  Just to show it has been this way for a while I used 3.5.4

- Git Extensions 3.5.4.12724
- Build 65f01f39982ccb121c4df7d4b00b506939553958
- Git 2.39.0.windows.1
- Microsoft Windows NT 10.0.19045.0
- .NET Framework 4.8.9105.0
- DPI 96dpi (no scaling)

Also take a look at https://git-scm.com/docs/git-submodule#Documentation/git-submodule.txt-absorbgitdirs  So if you fixed it by adding the correct .gitmodules file data and then ran that command, you would fix it to be a submodule.

**Reply by andry81** [reply]
> You have a funky state and I doubt is supported as valid git state.
> 
> So just because git is detecting nested repo and helping you out in summary call, it is not a submodule. The modules file is required to actually work git submodules.

It depends on the git and it's issues. The git already has the submodules state in the vcs database (for example, see the commit: https://github.com/andry81-tests/gil-test1-A/commit/f5f753fed28282a7297428e2eb158f948f719711).

And this is a primary storage, so I think the `.gitmodules` is a secondary storage to just for ease synchronization with the database. You are using the secondary storage. I don't think this is correct way of implementation.

**Reply by vbjay** [reply]
Nested repos are not submodules. Plain and simple. They are different animals. You need to stop trying to simplify apples and oranges.

**Reply by vbjay** [reply]
The reason we have a .git file in a submodule is to say submodule is at x sha. Submodule file holds info on how to get the needed sub repos. So your nested repo lacks the idea in master repo of what commit x repo is at.

 https://github.com/swcarpentry/git-novice/issues/272

**Reply by vbjay** [reply]
Read up on 

- nested repos
- submodules
- subtrees

**Comment by vbjay** [comment]
1. Add a .gitmodules file 
```
[submodule "_externals/gil-test1-B"]
	path = _externals/gil-test1-B
	url = https://github.com/andry81-tests/gil-test1-B.git
```
Before running
![image](https://user-images.githubusercontent.com/3629489/212490019-b58c43fd-49a7-4925-8a49-284a96fec99c.png)

2. Run the ```git submodule absorbgitdirs``` command
![image](https://user-images.githubusercontent.com/3629489/212490074-9a8c75a0-2954-4ea5-995e-d3aefdd44dd5.png)

After
![image](https://user-images.githubusercontent.com/3629489/212490086-9d86167b-9360-4b89-aaf3-635d1930a143.png)

So no...  .gitmodules is required to actually do submodule work.  Why we don't show it in modules since it technically is a nested repo and not a valid submodule state.

**Comment by andry81** [comment]
Just for instance. Have sent questions into git folks mailing list:
https://lore.kernel.org/git/1716310675.20230122233403@inbox.ru/

**Comment by andry81** [comment]
Seems [vcstool](https://github.com/dirk-thomas/vcstool) and the forks is the best choice here to avoid `.gitmodules` file in first place.

Discussions:
https://github.com/dirk-thomas/vcstool/issues/242

Forks:
https://github.com/aaronplusone/vcstool (https://github.com/aaronplusone/vcstool/tree/feature-sparse-checkouts)
https://github.com/MaxandreOgeret/vcstool2 (https://pypi.org/project/vcstool2/)

On another hand the `.gitmodules` file might be required in some circumstances and it's presence (and vcstool file too as well if hashes are used) in the default branch is not preferred, because will interfere with source files commits and may be left in the desync state because of a rewrite in a submodule, which will require a rewrite in all dependent repositories and basically is pain in arse. So better just push the `.gitmodules` file out of a default branch into standalone branch where you can rewrite without affecting the sources.

For example, use `master` branch to store sources without submodules and `master-modules` (or `master-all`) branch to store sources plus `.gitmodules` file to checkout default branch with head submodules or with `freezed` submodules. Later you can just rewrite `master-modules` without affecting the source files from `master` (there won't be merges from `master-modules` to `master`, only from `master` into `master-modules`).

This approach can use both the vcstool as a more convenient tool by default and the `.gitmodules` file, where it is might be required because of circumstances. For example, for a ZIP archive (`Download ZIP` button) as were noted [here](https://github.com/dear-github/dear-github/issues/214).
